File: Mio Language Design/Syntax Analyze Phase/CFG/LL1GeneralParser.py
class CFG:

    def __init__(self,start,cfg,selectionSet) -> None:

        self.index = 0
        self.word = []
        self.start = start
        self.cfg = cfg
        self.ss = selectionSet
        self.theEnd = False

    def reset(self):

        self.index = 0
        self.word = []
        self.theEnd = False

    def validate(self,word):

        self.word = word
        result = False

        if self.parser("<X>"):

            if len(word)-1 == self.index:
                result = True

        self.reset() #Reset parameters to parse more
        return result,word[:-1]

    def parser(self,start):

        accept = False
        rule = self.cfg[start]

        selectionSet = self.ss[start]
        if self.word[self.index] not in selectionSet:
            print(self.word[self.index],self.word[self.index] in selectionSet)
            return False

        for OR in rule:

            for k,terminal in enumerate(OR):

                accept = False
                if terminal[0] == '<': # if word is non-terminal

                    accept = self.parser(terminal) # Recursion starts

                    if not(accept): # Non Terminal failed to parse hence *break* to find other rule of same non terminal

                        if k>0: 
                            self.theEnd = True
                            #First Non Terminal return true this production rule is selected then second 
                            # or further Non Terminal returns false so no backtrack, Failed to parse

                        break

                elif terminal == self.word[self.index]: # If word match with Terminal in a rule

                    self.index += 1
                    accept = True

                else: # If word doesn't match with Terminal in a rule

                    # Word can be neglected if the terminal is null else break to find other rule of same Non terminal
                    if terminal == 'null': #return True instead break because it is end condition hence just returns True
                        if self.word[self.index] in selectionSet:
                            
                            return True
                        else:
                            
                            self.theEnd = True
                            return False

                    #First set of rule is true for word then second or further word false so no backtrack, Failed to parse
                    if k>0: 
                        self.theEnd = True
                    
                    break
            
            # This condition is on mean word wont parse so straightly Return False
            if self.theEnd:
                return False
                
            # this return True in order to skip searching in Next OR (production rule of same Non Terminal)
            # and it is also the here to give True if word parse correctly
            if accept:
                return True

            
        return False


"""
{ dt id = id } ;
{ dt id { dt id } } ;

<X> -> S ;
<S> -> { <A> }
<A> -> dt id <D> 
<D> -> <B> <C> | <S> | null
<B> -> = | += | *=
<C> -> intConst | floatConst
"""
selectionSet = {
            "<X>" : ['{','$'],

            "<S>" : ['{'],

            "<A>" : ['dt'],

            "<D>" : ["=", "+=", "*=",'{',"}"],

            "<B>" : ["=", "+=", "*="],

            "<C>" : ["intConst", "floatConst"]
}
A = CFG("<X>",
        {
            "<X>" : [   ["<S>",";"] ],

            "<S>" : [   ["{","<A>","}"] ],

            "<A>" : [   ["dt","id","<D>"] ],

            "<D>" : [   ["<B>","<C>"], ["<S>"], ["null"] ],

            "<B>" : [   ["="], ["+="], ["*="] ],

            "<C>" : [   ["intConst"], ["floatConst"] ]
        },
        selectionSet
    )
x= "{ dt id { dt id { dt id { dt id { dt id  } } }  } } ;"

print(A.validate(x.split()+["$"]))

x= "{ dt id  { dt id = intConst } } ;"

print(A.validate(x.split()+["$"]))



# A = CFG("<X>",
#         {
#             "<X>" : [ ["a","b"], ['c','d'] ]
#         }
#     )

# x= "c d"
# print(A.validate(x.split()+["$"]))

File: Mio/src/SyntaxAnalyzer/LL1GeneralParser.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package SyntaxAnalyzer;

import LexicalAnalyzer.TokenClass;
import java.util.HashMap;
import java.util.List;

/**
 *
 * @author Umar, Izhan, Muqsit
 */
public class LL1GeneralParser {
    
    String startTerminal;
    HashMap<String, String[][]> cfg;
    HashMap<String, String[]> ss;
    boolean theEnd = false;

    int index = 0;
    List <TokenClass> wordList;

    /**
     * 
     * @param start startTerminal token of grammar
     * @param contextFreeGrammar    context free grammar in the form of hash map
     * @param selectionSet  Selection Set of the grammar
     */
    public LL1GeneralParser(String start, 
                            HashMap<String, String[][]> contextFreeGrammar, 
                            HashMap<String, String[]> selectionSet) {
        
        this.startTerminal = start;
        this.cfg = contextFreeGrammar;
        this.ss = selectionSet;
    }
    
    /**
     * Reset index and token list so that new token can be parse
     * @return  void
     */
    private void reset(){
        this.index = 0;
        this.wordList = null;
        this.theEnd = false;
    }
    
    private boolean match(String terminal, int index, String token) {
        return String.valueOf(terminal.charAt(0)).equals(token);
    }
    
    private boolean match(String terminal, String token) {
        return terminal.equals(token);
    }
    
    private boolean match(String terminal, List <TokenClass> word) {
        String token = word.get(this.index).classP;
        return terminal.equals(token);
    }
    
    /**
     * 
     * @param tokens    ArrayList of Tokens or Word that are parse
     * @return result   true or false if tokens list matches with the grammar accordingly
     */
    public boolean validate(List <TokenClass> tokens) {
        this.wordList = tokens;
        boolean result = false;
        
        if ( this.parse(this.startTerminal) ) {
            if ( this.wordList.size() - 1 == this.index ){
                result = true;
            }
        }
        
        this.reset(); //Reset parameters to parse more
        
        return result;
    }
    
    private boolean parse(String nonTerminal){
        
        boolean accept = false;
        String [][] rule = this.cfg.get(nonTerminal); //Select rule
        
        for (String [] OR: rule) {
            
            for (int k = 0; k < OR.length; k++) {
                String token = OR[k];
                
                accept = false;
                
                // if token is non-terminal
                if (this.match(token, 0, "<")) 
                {
                    
                    accept = this.parse(token); // Recursion starts
                    
                    // Non Terminal failed to parse hence *break* to find other rule of same non terminal
                    if ( !(accept) ) {
                        if ( k > 0 ) {this.theEnd = true;}
                        break;
                    }
                }
                else if ( this.match(token, this.wordList) ) 
                {
                    this.index++;
                    accept = true;
                }
                else 
                {
                    if ( this.match(nonTerminal, "null") ) {
                        
                        return false;
                    }
                    
                    // First set of rule is true for word then second or further word false so no backtrack, Failed to parse
                    if ( k > 0 ) {this.theEnd = true;}
                    break;
                }
                
            }
        }
        return false;
    }
}
